### ISSUE TITLE
Frameless maximize overflow

### ISSUE THREAD

**Issue by adi518** [electron/electron#8728]
* Electron version: 1.4.15
* Operating system: Windows 10

### Actual behavior

Height is oversized by exactly 8px consistently (taskbar is set to small icons).

![Alt](https://raw.githubusercontent.com/adi518/electron-frameless-maximize-test/master/issue.jpg)

### How to reproduce

https://github.com/adi518/electron-frameless-maximize-test

**Comment by groundwater** [comment]
Hey @adi518. I'd like to help this issue get prioritized. Can you fill out a few more details about what's going wrong and how exactly to reproduce it?

**Comment by adi518** [comment]
All I did is set `frame` to `false` and hit maximize to test it. It's easily reproduced in my repo, which also has the devtools enabled.

**Comment by groundwater** [comment]
Hi @adi518. This is a large repo. Can you provide a minimal reproduction case? Thank you.

**Comment by adi518** [comment]
Sure, I'll get ya something.

**Comment by adi518** [comment]
Hi, made a minimal repro: https://github.com/adi518/electron-frameless-maximize-test

**Comment by radencode** [comment]
@adi518 @groundwater I am running into the same issue and I still am not able to find a solution for it.

**Comment by thesbros** [comment]
This seems to be caused by PR #8404. Reverting that PR (170f2f61f9be92ecf16dba54113d840730400199) fixes the page getting cut off.

cc/ @bsclifton

**Comment by radencode** [comment]
@thesbros does that work for you?

**Comment by bsclifton** [comment]
@thesbros the change in https://github.com/electron/electron/pull/8404 is needed for > 100% DPI because otherwise the coordinates aren't scaled properly. If reverted, you'll see a white or gray borders at 200% DPI

I believe the root cause for the issue reported here is captured with https://github.com/electron/electron/issues/4045 (specifically, a similar issue was reported with https://github.com/electron/electron/issues/8546 which got merged w/ 4045).

Can anyone try the demo program on Windows 7? My guess is that the size is calculated correctly there

**Comment by adi518** [comment]
Actually, I gave it to a friend the other day, who happens to run Win7 and he experienced the same issue.

**Comment by thesbros** [comment]
@bsclifton I understand what the fix is needed for - but arguably 100% DPI is the most common setup. Is this just blocked on #4045 then?

I can spin up a W7 VM later, but @adi518 says it still exists on 7.

**Comment by bsclifton** [comment]
@thesbros I don't see any comment by @adi518 about problem existing on Win7.  However, I tried it on my VM and confirmed the issue exists
![screen shot 2017-02-27 at 2 01 34 am](https://cloud.githubusercontent.com/assets/4733304/23355000/e4fe5132-fc90-11e6-9902-4b3213d27aae.png)

**Comment by adi518** [comment]
Yes, it definitely happens with Win7.

**Comment by radencode** [comment]
@bsclifton @adi518 @thesbros So is there a fix for the oversized window on maximized state? I've been playing around with it for a while and can't find a solution for it.

**Comment by adi518** [comment]
Unfortunately, nothing yet. Apparently this issue was caused by a commit fixing another issue, some weeks ago. I hope @bsclifton could come up with something, otherwise Electron frame-less functionality is pretty much broken for everyone.

**Comment by radencode** [comment]
Yeah I guess we wait and try to figure it out.

**Comment by bsclifton** [comment]
@adi518 @radarhere Part of the problem is that when a window is maximized, Windows sets the X/Y to the negative frame width/height.  For example, I believe on Windows 10, it was -7/-7.

One possible work-around, which I [captured here](https://github.com/electron/electron/issues/5267#issuecomment-272607225) would be to manually override the min/max properties.  An alternative to that would be to put a handler on the WM_WINDOWPOSCHANGED and manually set the x/y to 0,0

Reverting https://github.com/electron/electron/pull/8404 would be a hack.  The reason that it seems to make the alignment work properly for the original post is because it forces the bounds frame (even though this is a frameless window).  Frameless windows should not have frames, IMO.  When you dock the taskbar to the top, the overflow works the same; the overflowed content goes to the bottom and bleeds under the taskbar (as seen with https://github.com/electron/electron/issues/8546)

**Comment by bsclifton** [comment]
BTW: Spy++ is a great tool to use for debugging- you can find the HWND and see the bounds yourself

**Comment by radencode** [comment]
@bsclifton Yes but when we override the bounds to (0,0) then we get a white border around our frameless window and to get rid of that we need to remove the styles from thickframe which causes all of the animations to be lost. Therefore we are stuck at a dead end?!

**Comment by thesbros** [comment]
@bsclifton I did the `MINMAXINFO` override in #8782 - but that only fixes #5267, not this issue.

**Comment by bsclifton** [comment]
@radencode can you share a screenshot of the white borders? this sounds like the real issue

**Comment by radencode** [comment]
@bsclifton Yes I can do that but I am currently at work. I will share it as
soon as I get home.

**Comment by radencode** [comment]
![border2](https://cloud.githubusercontent.com/assets/25285792/23391604/16674284-fd2b-11e6-8012-c31616785579.jpg)
![border](https://cloud.githubusercontent.com/assets/25285792/23391605/1a311ffc-fd2b-11e6-9a0c-9a9f79da144e.jpg)

@bsclifton Here is the screen shot of the white border. I no longer have the code because it did not fix my issue and I overrode it with other tests. @thesbros tested this example as well so he might be able to provide the code of what we did. Also I am not very familiar with the `native_window_views.cc` code and I may have missed a step. Can you please provide more details of how you override the bounds (x,y) and also does it work for you, have you tested it?

**Comment by bsclifton** [comment]
@radencode here's an example you can try to get more info (I don't have electron compiled and ready to test at the moment unfortunately and it's already 2am)

```diff
diff --git a/atom/browser/native_window_views_win.cc b/atom/browser/native_window_views_win.cc
index 1b523e9..100f370 100644
--- a/atom/browser/native_window_views_win.cc
+++ b/atom/browser/native_window_views_win.cc
@@ -5,6 +5,7 @@
 #include "atom/browser/browser.h"
 #include "atom/browser/native_window_views.h"
 #include "content/public/browser/browser_accessibility_state.h"
+#include <cmath>

 namespace atom {

@@ -147,6 +148,51 @@ bool NativeWindowViews::PreHandleMSG(
       }
       return false;
     }
+
+    //\see WM_GETMINMAXINFO - https://msdn.microsoft.com/en-us/library/windows/desktop/ms632626(v=vs.85).aspx
+    case WM_GETMINMAXINFO: {
+      MINMAXINFO* mmi = (MINMAXINFO*)l_param;
+
+      if (IsMaximized()) {
+        HWND hwnd = GetAcceleratedWidget();
+        RECT window_rect;
+        if (!GetWindowRect(hwnd, &window_rect)) {
+          return false;
+        }
+
+        HMONITOR monitor = MonitorFromRect(&window_rect, MONITOR_DEFAULTTONULL);
+        if (!monitor) {
+          return false;
+        }
+
+        MONITORINFO monitor_info = { 0 };
+        monitor_info.cbSize = sizeof(monitor_info);
+        GetMonitorInfo(monitor, &monitor_info);
+
+        RECT work_area = monitor_info.rcWork;
+        RECT monitor_rect = monitor_info.rcMonitor;
+
+        mmi->ptMaxPosition.x = std::abs(work_area.left - monitor_rect.left);
+        mmi->ptMaxPosition.y = std::abs(work_area.top - monitor_rect.top);
+        mmi->ptMaxSize.x = std::abs(work_area.right - work_area.left);
+        mmi->ptMaxSize.y = std::abs(work_area.bottom - work_area.top);
+        mmi->ptMinTrackSize.x = 2;
+        mmi->ptMinTrackSize.y = 2;
+
+        *result = 0;
+        return true;
+      }
+
+    //\see WM_WINDOWPOSCHANGING - https://msdn.microsoft.com/en-us/library/windows/desktop/ms632653(v=vs.85).aspx
+    case WM_WINDOWPOSCHANGING: {
+      WINDOWPOS* wp = (WINDOWPOS*)l_param;
+      if (IsMaximized()) {
+        LOG(ERROR) << "WM_WINDOWPOSCHANGING; position=[" << wp->x << "," << wp->y <<
+        "]; width=[" << wp->cx << "," << wp->cy;
+        return false;
+      }
+    }
+
     default:
       return false;
   }
```

In `WM_GETMINMAXINFO`, I used the same calculations which Chromium uses so this should be accurate (see `ui/views/win/hwnd_message_handler.cc` in Chromium 54 or newer)

In `WM_WINDOWPOSCHANGING`, I just show how you can log the params. This should help with understanding what the x/y and width/height are set to. Because of the borders, I believe these values are incorrect.  If you wanted to adjust these, you can set the `x`/`y` and `cx`/`cy` for the `wp` variable, set *result = 0, and return true.

**Comment by thesbros** [comment]
@bsclifton It fixes the monitor overflow issue for me - but still not this issue. Does it require changing the x/y/cx/cy values to fix that?

**Comment by bsclifton** [comment]
@thesbros very likely it may- needs to be tested.  I might have time later today

**Comment by tonyganch** [comment]
Hi guys, I think this issue is a duplicate of #8454.
Copy-pasting my comment from there just in case:

I've git-bisected the repo to see that the line causing the issue is this one:
https://github.com/electron/electron/blob/master/atom/browser/ui/win/atom_desktop_window_tree_host_win.cc#L33

When I remove that, windows with both `frame: false` and `frame: true` behave okay on my Windows 10 machine.
I see that it's been added in this commit (9e0547b) as part of this PR (#8404).
Can someone please explain why we need that method/line?

**Comment by bsclifton** [comment]
@tonyganch you're welcome to submit a PR which reverts https://github.com/electron/electron/commit/9e0547b98f8b7f3a572aa1cf9770f434c8fdeed9

I work with Brave and we maintain a fork of Electron.  This patch was required for us starting with Chromium 54 to get rid of a solid white ~7px border around all edges of the screen for folks with larger than 100% DPI.  By returning an inset of 0, you're telling Chromium the client area is the same size as the window (which is true for frameless windows).  After we successfully released (our fork does not see this issue), I upstreamed the patch

At this point, I would recommend reverting the above commit if folks can test and ensure it looks correct when at 125%, 150% and 200% DPI

**Comment by thesbros** [comment]
@adi518 Would you mind trying out [this build?](https://github.com/thesbros/electron/releases/download/v1.6.1/R.zip)

@bsclifton I fixed the issue with [this patch](https://github.com/thesbros/electron/commit/f0ea66582ac4ebeae240660a4c60ce63f2464d18), but it causes the edges of the page to be cut off with DPI > 100%. Could you give any pointers on how to fix that?

**Comment by adi518** [comment]
@thesbros I'm glad to report your build fixes this issue.

**Comment by thesbros** [comment]
@adi518 Great! I will be submitting a PR as soon as I fix the DPI issue.

**Comment by adi518** [comment]
I'm afraid it's not over yet. I was testing the last fix on my machine at work, where the taskbar was set to bottom. At home I have it set to top and although the height is now correct, I get a thick top-border above the window.

![Alt](https://raw.githubusercontent.com/adi518/electron-frameless-maximize-test/master/issue-after-@thesbros-fix-attempt.jpg)

**Comment by thesbros** [comment]
@adi518 Have you tried my latest commit in #8808? I can't reproduce this issue with my PR.

**Comment by adi518** [comment]
I think I can build it. I'll let you know soon. Thanks.

**Comment by adi518** [comment]
Ok, forgot I don't have VSCE on this machine, so that will take some time to setup. Upload a build if you may, thanks.

**Comment by thesbros** [comment]
@adi518 [Here you go.](https://github.com/thesbros/electron/releases/download/v1.6.1/R.zip)

**Comment by adi518** [comment]
Sweet, the border is gone! taskbar at top/bottom now appears to works correctly.

**Comment by adi518** [comment]
Oh wait a minute. My app is dark-grey, I now noticed it actually has round corners at the top. Bottom is still rectangular though.

**Comment by thesbros** [comment]
@adi518 Yes the rounded edges are a side effect. I am currently trying to figure out how to get rid of them.

**Comment by adi518** [comment]
Haha. Yea, now I recall reading a comment about it in the PR thread.

**Comment by Venipa** [comment]
now its overflowing again in full screen if you are using browserviews:
![image](https://user-images.githubusercontent.com/17952364/116165955-391bec00-a6fd-11eb-8180-4baed6633d39.png)
it works correctly if not fullscreen:
![image](https://user-images.githubusercontent.com/17952364/116166006-594bab00-a6fd-11eb-81b2-c7b9b9f6b4c3.png)

**Comment by Awendel** [comment]
> now its overflowing again in full screen if you are using browserviews: ![image](https://user-images.githubusercontent.com/17952364/116165955-391bec00-a6fd-11eb-8180-4baed6633d39.png) it works correctly if not fullscreen: ![image](https://user-images.githubusercontent.com/17952364/116166006-594bab00-a6fd-11eb-81b2-c7b9b9f6b4c3.png)

Can confirm the exact issue 100%.

Any fixes for BrowserView?

Otherwise will have to try to manually adjust
